docs/error-handling.md

```diff
@@ -0,0 +1,437 @@
+# Error Handling
+
+## Basics
+
+Functions that may fail, and return a pointer, generally look like this:
+```c
+/* Create a new thing.
+ * Returns NULL if creation failed, use aws_last_error() to get the error code explaining why. */
+struct aws_thing *aws_thing_new(...);
+```
+
+Functions that may fail, but don't return a pointer, generally look like this:
+```c
+/* Return AWS_OP_SUCCESS (0) if thing successfully did something.
+ * Returns AWS_OP_ERR (-1) if it failed, use aws_last_error() to get the error code explaining why. */
+int aws_thing_do_something(...);
+```
+
+`aws_last_error()` returns an error code that's stored in a thread-local variable. Any function that fails will make sure to set that thread-local error code by calling `aws_raise_error(int error_code)`.
```

Comment:
should we note somewhere here that even though return is int, the only 2 valid values are SUCCESS and ERR?

Reply:
i see it mentioned later, but i dont see explanation for why we do it this way

Reply:
I'll say it again here. You're right, it is very important to understand